src/NetscriptPort.ts

```diff
@@ -76,9 +79,14 @@ export function peekPort(n: PortNumber): PortData {
   return port.data[0];
 }
 
-function nextWritePort(n: PortNumber) {
+function nextWritePort(n: PortNumber, ctx: NetscriptContext) {
   const { resolvers } = getPort(n);
-  return new Promise<void>((res) => resolvers.push(res as Resolver));
+  return new Promise<void>((res, rej) => resolvers.push(() => (ctx.workerScript.env.stopFlag ? rej() : res()))).catch(
+    () => {
+      helpers.log(ctx, () => `nextWrite was rejected on port ${n}`);
+      throw new ScriptDeath(ctx.workerScript);
+    },
+  );
 }
```

Comment:
This could also be done with an `AbortSignal` to completely clear the reference and abort the promise early, so that it doesn't wait until the next time the port is written to. The other benefit of allowing the user to pass in a signal, is that they can control the lifetime of when they want to work with the event.

```ts
function nextWritePort(n: PortNumber, signal: AbortSignal) {
  const { resolvers } = getPort(n);
  return new Promise<void>((res, rej) => {
    resolvers.push(res as Resolver);
    signal.addEventListener(
      "abort",
      () => {
        rej(signal.reason);
        const resolverIndex = resolvers.indexOf(res);
        if (resolverIndex !== -1) resolvers.splice(resolverIndex, 1);
      },
      { once: true }
    );
  });
}
```

It could be expanded using the context and an AbortSignal on the context to reject the promise as well so that it would reject on the script ending. - It would help with stopping other async functionalities on script end as well.